implemented accounts data cache

pompon0 opened the pull request:
https://nearinc.atlassian.net/browse/CPR-94

accounts_data::Cache is the central component collecting the SignedAccountData (data about validator nodes broadcasted by the the validators themselves). It takes care of preventing DDOS attacks of various types: minimizes the amount of signature verifications (which is computationally expensive) and detects malicious data as early as possible. It will be plugged into the PeerManager logic in the next PR.

matklad commented:
Could you include a short summary in the PR description of what is this for? Opening the jira link, the design document, and ctrl+F ing for "cache" didn't turn out any results. 

If it is already hard to understand the context when the PR is submitted, it would be almost impossible to do so several years down the line when someone stumbles upon this PR in git blame. 

This also can benefit from a rebase, now that #7206 is in.

matklad commented on chain/network/src/accounts_data/mod.rs:
```diff
@@ -0,0 +1,240 @@
+        drop(epochs);
+
+        // We verify signatures synchronously for now.
+        // To verify signatures in parallel, we should have a way to stop verification at the first invalid one.
```
https://docs.rs/rayon/latest/rayon/iter/trait.ParallelIterator.html#method.find_any

pompon0 replied:
I need the partial result as well. find_any won't help.

matklad commented on chain/network/src/accounts_data/mod.rs:
```diff
@@ -0,0 +1,240 @@
+            }
+            data.push(d);
+        }
+        return (data, None);
```
```suggestion
        (data, None)
```

I feel like I make redundant return suggestion fairly frequently, please try to stick to idiomatic rust :)

pompon0 replied:
I'm not doing that on purpose, and I think that usually I get it right (just not always). The difference is marginal, so I need to really concentrate to spot things like that. If that's idiomatic, maybe "cargo fmt" should do those fixes?

matklad replied:
`cargo fmt` doesn't do this as this feature needs a bit of control-flow analysis to detect expression in tail positions. But I believe `cargo clippy` would complain, there's an on-by-default lint for that: 

https://rust-lang.github.io/rust-clippy/master/#needless_return

pompon0 commented on chain/network/src/accounts_data/mod.rs:
```diff
@@ -0,0 +1,240 @@
+        let x = self.clone();
+        // Execute insertion in a dedicated runtime.
+        self.runtime
+            .spawn(async move {
```
Are you sure that it is a good idea to have all CPU-heavy tasks on a single thread pool? Is there no threat of starvation in such setup?

pompon0 replied:
> dedicated threadpools feel like a work-around
> oversubscribing CPU also causes starvation, just in a different way.

Starvation is when you don't have a guarantee that a given task will eventually complete because it won't even obtain enough resources. A bounded number of threadpools (with bounded number of threads) with a round robin CPU assignment precisely gives a guarantee that no threadpool will starve.

> for known cases where we are spawning unbounded amount of background work, implement throttling, like we do here

This specific implementation of throttling under-utilizes CPU as in case there is no competing tasks we utilize only half of the CPUs. Also it is not so different than just assigning a dedicated threadpool with rayon::current_num_threads to the given task.

@matklad how is this rayon example really different than spawning exactly the same stuff on tokio, except that it doesn't support async waiting for results but rather makes you use extra channels?

@mm-near what is your thoughts on sharing single global thread pool for block processing and and other stuff? Didn't we have already issues of this sort?

pompon0 replied:
yes: 
https://github.com/matklad/nearcore/blob/5c582ce9bba19a9d74bf6858179c381d78647882/chain/chain/src/chain.rs#L1967-L1998

pompon0 commented:
Added purpose of accounts_data::Cache to the PR description.
Rebased to master.
I've also changed the API of accounts_data::Cache to simplify it in the context of the changes which I made in the next PR.

Before I jump to using rayon, I have a couple of questions:
1. As I understand, rayon (globally) uses a threadpool of fixed size and since closures spawned to rayon are blocking (i.e. not async) this means that if 2 callers use the global threadpool at the same time, they can starve each other. I don't think it would be a good idea, for example, for block processing to be (potentially significantly) delayed by some PeerManager processing (they share the CPU, but the simple round robin enforced by the OS would at least prevent starvation). Don't you agree?
2. I want to asynchronously wait for the verification to finish, so that the result can be stored and broadcasted. rayon doesn't have any async support so doing it "rayon way" would require to add synchronization primitives on top of rayon stuff. Is it really worth it? To me is more extra complexity than gain, really.

matklad commented on chain/network/src/accounts_data/mod.rs:
```diff
@@ -0,0 +1,204 @@
+}
+
+impl CacheInner {
+    fn is_new(&self, d: &SignedAccountData) -> bool {
+        let id = (d.epoch_id.clone(), d.account_id.clone());
+        self.keys.contains_key(&id)
+            && match self.data.get(&id) {
+                Some(old) if old.timestamp >= d.timestamp => false,
+                _ => true,
+            }
+    }
+    fn try_insert(&mut self, d: SignedAccountData) -> Option<SignedAccountData> {
```
I wonder if there's some lightweight way to distinguish potentially unverifid data in the first-case from the verified data in the second case. Probably isn't worth it though.

pompon0 replied:
I'm afraid I do not follow your concern here.

matklad commented on chain/network/src/accounts_data/tests.rs:
```diff
@@ -0,0 +1,207 @@
+        cache.clone().insert(vec![a0.clone(), a1.clone(), a2old.clone(), a2new.clone()]).await;
+    assert_eq!(Some(Error::SingleAccountMultipleData), res.1);
+    // Partial update is allowed, in case an error is encountered.
+    assert!([&a0, &a1, &a2old, &a2new].as_set().is_superset(&res.0.as_set()));
```
Rather than adding `AsSet` trait, I think it'll be more valuable to add an `#[track_caller] fn assert_super_set` function to emit a useful error message her.

pompon0 replied:
I assume that you are fine with assert_eq(x.as_set(),y.as_set()) and the concern is about just the error message for is_superset. I'll try to find a way to pretty print the diff for that cases.

pompon0 replied:
done

matklad commented on chain/network-primitives/src/types.rs:
```diff
@@ -319,6 +321,61 @@ pub enum NetworkViewClientMessages {
+
+impl<T> std::cmp::PartialEq for WithHash<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.hash == other.hash
```
This... feels potentially wrong? hash very much does not guarantee that different values hash to different u64s, especially given that we don't randomize the hash seed. This is also confusing, as I assumed that `WithHash` is just for caching the hash, and almost missed overridden `eq`.

pompon0 replied:
Whole blockchain technology relies on hashes not colliding :P. What if I put here a sha256 instead? I'm using a weaker hash here because there is no adversary. Also the number of pairs of values compared is order of magnitude smaller than what is usually required for hashing functions. To make it a truly correct probabilisitic algo, I could select a specific hashing algorithm (rather than whatever hashmap uses) with decent properties and seed it at random on binary startup. Other alternatives that I see are as follows:
1. Go back to identifying sets with EpochId and have an unenforced invariant that EpochId determines the set.
2. Just always do a full comparison and have an unenforced invariant that ClientActor won't send us updates more often than necessary.

pompon0 replied:
Then Arc::ptr_eq it is.

matklad reviewed (commented):
I think this is broadly OK, but I need more convincing for:

* the way we do background work -- using rayon still feels more appropriate to me
* WithHash shenanigans -- relying on Hash for Eq feels wrong.

pompon0 commented on chain/network/src/accounts_data/mod.rs:
```diff
@@ -0,0 +1,204 @@
+        self: Arc<Self>,
+        data: Vec<SignedAccountData>,
+    ) -> (Vec<SignedAccountData>, Option<Error>) {
+        let x = self.clone();
```
done

pompon0 commented on chain/network-primitives/src/types.rs:
```diff
@@ -319,6 +321,61 @@ pub enum NetworkViewClientMessages {
 }
 
+#[derive(Debug)]
+pub struct WithHash<T> {
+    inner: T,
+    hash: u64,
+}
+
+impl<T: Hash> Hash for WithHash<T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_u64(self.hash);
+    }
+}
+
+impl<T: Hash + Default> Default for WithHash<T> {
+    fn default() -> Self {
+        WithHash::new(T::default())
+    }
+}
+
+impl<T> std::cmp::PartialEq for WithHash<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.hash == other.hash
+    }
+}
+
+impl<T> std::ops::Deref for WithHash<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T: Hash> WithHash<T> {
+    pub fn new(inner: T) -> Self {
+        let mut hasher = std::collections::hash_map::DefaultHasher::new();
+        inner.hash(&mut hasher);
+        Self { inner, hash: hasher.finish() }
+    }
+}
```
removed

matklad commented on chain/network/src/accounts_data/mod.rs:
```diff
@@ -0,0 +1,206 @@
+
+async fn rayon_spawn<T: 'static + Send>(f: impl 'static + Send + FnOnce() -> T) -> T {
+    let (send, recv) = tokio::sync::oneshot::channel();
+    rayon::spawn(move || send.send(f()).ok().unwrap());
```
Hm, I'd `let _` instead of `unwrap` here. Futures are cancellable, and it's hard to rule out the case where something somewhere above the stack ends up in a `select` or `timeout`. 

To be more specific: I think putting a future which is not explicitly prepated for cancellation into select is the mistake, but its an easy mistake to make, so I'd love us to sort-of defensively code in this situation: just wasting the time computing `f` seems better than crashing the whole process.

pompon0 replied:
Added a must_complete wrapper for futures which are expected to complete.
If an incomplete future is dropped, the wrapper calls log_assert (which panics in tests, and logs an error in prod). Also added a log_assert wrapping the send.send().is_ok() call.

matklad reviewed (approved):
LGTM, though a mild preference for replacing `.unwrap` with ignoring a send error here: https://github.com/near/nearcore/pull/7208/files#r927798784

pompon0 commented on chain/network/src/accounts_data/mod.rs:
```diff
@@ -0,0 +1,206 @@
+    pub fn set_keys(&self, keys: Arc<AccountKeys>) -> bool {
+        let mut inner = self.inner.write();
+        // Skip further processing if the key set didn't change.
+        if Arc::ptr_eq(&keys, &inner.keys) || keys == inner.keys {
```
ack
